packages/@tailwindcss-upgrade/src/migrate-js-config.ts

```diff
@@ -0,0 +1,185 @@
+    cssConfigs.push(migrateDarkMode(unresolvedConfig as any))
+  }
+
+  if ('content' in unresolvedConfig) {
+    sources = migrateContent(unresolvedConfig as any, base)
+  }
+
+  if ('theme' in unresolvedConfig) {
+    cssConfigs.push(await migrateTheme(unresolvedConfig as any))
+  }
+
+  return {
+    sources,
+    css: cssConfigs.join('\n'),
+  }
+}
+
+async function migrateTheme(unresolvedConfig: Config & { theme: any }): Promise<string> {
+  let { extend: extendTheme, ...overwriteTheme } = unresolvedConfig.theme
+
+  let resetNamespaces = new Map<string, boolean>()
+  // Before we merge the resetting theme values with the `extend` values, we
+  // capture all namespaces that need to be reset
+  for (let [key, value] of themeableValues(overwriteTheme)) {
+    if (typeof value !== 'string' && typeof value !== 'number') {
+      continue
+    }
+
+    if (!resetNamespaces.has(key[0])) {
+      resetNamespaces.set(key[0], false)
+    }
+  }
+
+  let themeValues = deepMerge({}, [overwriteTheme, extendTheme], mergeThemeExtension)
+
+  let prevSectionKey = ''
+
+  let css = `@theme {`
+  for (let [key, value] of themeableValues(themeValues)) {
+    if (typeof value !== 'string' && typeof value !== 'number') {
+      continue
+    }
+
+    let sectionKey = createSectionKey(key)
+    if (sectionKey !== prevSectionKey) {
+      css += `\n`
+      prevSectionKey = sectionKey
+    }
+
+    if (resetNamespaces.has(key[0]) && resetNamespaces.get(key[0]) === false) {
+      resetNamespaces.set(key[0], true)
+      css += `  --${keyPathToCssProperty([key[0]])}-*: initial;\n`
+    }
+
+    css += `  --${keyPathToCssProperty(key)}: ${value};\n`
+  }
+
+  return css + '}\n'
+}
+
+function migrateDarkMode(unresolvedConfig: Config & { darkMode: any }): string {
+  let variant: string = ''
+  let addVariant = (_name: string, _variant: string) => (variant = _variant)
+  let config = () => unresolvedConfig.darkMode
+  darkModePlugin({ config, addVariant })
+
+  if (variant === '') {
+    return ''
+  }
+  return `@variant dark (${variant});\n`
+}
+
+// Returns a string identifier used to section theme declarations
+function createSectionKey(key: string[]): string {
+  let sectionSegments = []
+  for (let i = 0; i < key.length - 1; i++) {
+    let segment = key[i]
+    // ignore tuples
+    if (key[i + 1][0] === '-') {
+      break
+    }
+    sectionSegments.push(segment)
+  }
+  return sectionSegments.join('-')
+}
+
+function migrateContent(
+  unresolvedConfig: Config & { content: any },
+  base: string,
+): { base: string; pattern: string }[] {
+  let sources = []
+  for (let content of unresolvedConfig.content) {
+    if (typeof content !== 'string') {
+      throw new Error('Unsupported content value: ' + content)
+    }
+    sources.push({ base, pattern: content })
+  }
+  return sources
+}
+
+// Applies heuristics to determine if we can attempt to migrate the config
+function isSimpleConfig(unresolvedConfig: Config, source: string): boolean {
```

Comment:
Should we rename this to something that will remain accurate even as we support more complex configs? Like `canMigrateConfig` or something?